### PR TITLE
Authenticating via Discord

### PR DESCRIPTION
**Pull request by robdy**
Todo list:
- [x] Authenticating to Discord
- [x] Obtaining user consent and data
- [x] Adding user to the server
- [x] Changing nickname
- [x] Adding appropriate role
- [x] Logging info FHQ log

Must have:
- [x] Proper error catching (I sometimes receive unhandledException error) - done thx to @pkmnbumba 
- [x] Add rate limit support

Good to have:
- [ ] (?) Removing the association by moderator
- [ ] (?) Disassociating the account/changing the associated account
- [ ] Add state checking during the auth (we don't have it on Reddit.js so is it really needed?)
- [ ] Add some nice frontend
- [ ] Finish this todo list with other stuff
- [ ] Saving account info into user profile and displaying it (not really must if we have logging)

To test: add the Discord bot data and go to http://localhost:1337/discord. It'll return your account info to your browser.

### CODE REVIEW

**Comment by robdy** [comment]
Added the most important file which I forgot yesterday. Also, I added joining to the server but it's not working for some reasons :thinking:

**Comment by robdy** [comment]
I think it's ready for testing :eyes:

**Comment by pkmnbumba** [comment]
Commenting to note that we discussed having an additional check to see if users have flair set before allowing them to join. Will try to work on it this weekend.

File: api/controllers/AuthController.js [hunk 1]
```diff
@@ -92,5 +92,39 @@ module.exports = {
...
+      const response = await Discord.getAccessToken(code);
+      const accessToken = response.access_token;
+      const currentUser = await Discord.getCurrentUser(accessToken);
+      const nick = req.user.name;
+      const joinedUser = await Discord.addUserToGuild(accessToken, currentUser.id, nick);
+      const serverUrl = 'https://discordapp.com/channels/' + sails.config.discord.server_id;
+      if (!joinedUser) {
+        return res.redirect(serverUrl);
+      }
+      sails.log(req.user.name +
```
Comment (pkmnbumba) [inline api/controllers/AuthController.js]: Looks like we need to add logging back here.

`await Event.create({type: "discordJoin", user: nick,content: "Joined Discord as @" + currentUser.username + "#" + currentUser.discriminator + " (ID: " + currentUser.id + ")"});`

File: api/services/Discord.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    return currentUser;
+  } catch (err) {
+    throw {error: 'Error retrieving current user from Discord; Discord responded with status code ' + err.statusCode};
+  }
+};
+
+exports.addUserToGuild = async function (token, user, nick) {
+  const route = 'https://discordapp.com/api/guilds/' + sails.config.discord.server_id + '/members';
+  const url =  route + '/' + user;
+  const auth = 'Bot ' + sails.config.discord.client_token;
```
Comment (pkmnbumba) [inline api/services/Discord.js]: `sails.config.discord.client_token` should be `sails.config.discord.bot_token` based on your local example

**Comment by robdy** [comment]
Tests I did:

- [x] from non-admin user (with flair set) joined successfully. Log entry created
- [x] same user, tried to join again - redirected to server
- [x] cleared flair and record from tables `user` - didn't see the button + when I tried go to `/discord` I got 403
- [x] set flair again and then joined successfully

File: api/controllers/AuthController.js (hunk 1 above)
Comment (robdy) [reply api/controllers/AuthController.js]: Done

File: api/services/Discord.js (hunk 2 above)
Comment (robdy) [reply api/services/Discord.js]: Done

**Pull request merged**